pkg/runtime/pubsub/bulkpub.go

```diff
@@ -0,0 +1,111 @@
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pubsub
+
+import (
+	"sync"
+
+	contribPubsub "github.com/dapr/components-contrib/pubsub"
+)
+
+const (
+	bulkPublishSeriallyKey string = "bulkPublishSerially"
+)
+
+// defaultBulkPublisher is the default implementation of BulkPublisher.
+// It is used when the component does not implement BulkPublisher.
+type defaultBulkPublisher struct {
+	p contribPubsub.PubSub
+}
+
+// NewDefaultBulkPublisher returns a new defaultBulkPublisher from a PubSub.
+func NewDefaultBulkPublisher(p contribPubsub.PubSub) *defaultBulkPublisher {
+	return &defaultBulkPublisher{
+		p: p,
+	}
+}
+
+// BulkPublish publishes a list of messages to a topic as individual Publish requests.
+// If 'bulkPublishSerially' metadata is set to true, the messages are sent to the broker serially,
+// in the same order. Otherwise they are sent to the broker as parallel Publish requests.
+func (p *defaultBulkPublisher) BulkPublish(req *contribPubsub.BulkPublishRequest) (contribPubsub.BulkPublishResponse, error) {
+	if req.Metadata[bulkPublishSeriallyKey] == "true" {
+		return p.bulkPublishSerial(req)
+	} else {
+		return p.bulkPublishParallel(req)
+	}
+}
+
+// bulkPublishSerial publishes messages in a serial order. This is slower, but ensures
+// that messages are sent to the broker in the same order as specified in the request.
+func (p *defaultBulkPublisher) bulkPublishSerial(req *contribPubsub.BulkPublishRequest) (contribPubsub.BulkPublishResponse, error) {
+	statuses := make([]contribPubsub.BulkPublishResponseEntry, 0, len(req.Entries))
+
+	for _, entry := range req.Entries {
+		statuses = append(statuses, p.bulkPublishSingleEntry(req, entry))
+	}
+
+	return contribPubsub.BulkPublishResponse{Statuses: statuses}, nil
+}
+
+// bulkPublishParallel publishes messages in parallel. This is faster, but does not guarantee
+// that messages are sent to the broker in the same order as specified in the request.
+func (p *defaultBulkPublisher) bulkPublishParallel(req *contribPubsub.BulkPublishRequest) (contribPubsub.BulkPublishResponse, error) {
+	statuses := make([]contribPubsub.BulkPublishResponseEntry, 0, len(req.Entries))
+	var wg sync.WaitGroup
+
+	statusChan := make(chan contribPubsub.BulkPublishResponseEntry, len(req.Entries))
+
+	for _, entry := range req.Entries {
+		wg.Add(1)
+
+		go func(entry contribPubsub.BulkMessageEntry) {
+			defer wg.Done()
+			statusChan <- p.bulkPublishSingleEntry(req, entry)
+		}(entry)
+	}
+
+	wg.Wait()
+	close(statusChan)
+
+	for status := range statusChan {
+		statuses = append(statuses, status)
```

Comment:
same as above ? since mem is already allocated. 
```suggestion
		statuses[i] = status
```

Reply:
Same as above, `statuses := make([]contribPubsub.BulkPublishResponseEntry, 0, len(req.Entries))`

Reply:
Same as above, since capacity is already allocated, it should be okay, what do you think?

"If the capacity of s is not large enough to fit the additional values, append allocates a new, sufficiently large underlying array that fits both the existing slice elements and the additional values. Otherwise, append re-uses the underlying array."
https://go.dev/ref/spec#Appending_and_copying_slices